lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java

```diff
@@ -1221,6 +1228,110 @@ private Location createListingLocation(String path)
         return createLocation(path);
     }
 
+    @Test
+    void testFileDoesNotExistUntilClosed()
+            throws Exception
+    {
+        Location location = getRootLocation().appendPath("testFile2");
+        getFileSystem().deleteFile(location);
+        try (OutputStream out = getFileSystem().newOutputFile(location).create()) {
+            assertFalse(fileExistsInListing(location));
+            assertFalse(fileExists(location));
+            out.write("test".getBytes(UTF_8));
+            assertFalse(fileExistsInListing(location));
+            assertFalse(fileExists(location));
+        }
+        assertTrue(fileExistsInListing(location));
+        assertTrue(fileExists(location));
+        getFileSystem().deleteFile(location);
+    }
+
+    @Test
+    public void testLargeFileDoesNotExistUntilClosed()
+            throws IOException
+    {
+        Location location = getRootLocation().appendPath("testFile3");
+        getFileSystem().deleteFile(location);
+        byte[] bytes = new byte[8192];
+        for (int i = 0; i < bytes.length / 4; i++) {
+            //noinspection NumericCastThatLosesPrecision
+            bytes[i] = (byte) i;
+        }
+        try (OutputStream outputStream = getFileSystem().newOutputFile(location).create()) {
+            // write a 32 mb file
+            for (int i = 0; i < divide(32 * 1024 * 1024, bytes.length, UNNECESSARY); i++) {
+                outputStream.write(bytes);
+                if (i + bytes.length >= 32 * 1024 * 1024) {
+                    assertFalse(fileExistsInListing(location));
+                    assertFalse(fileExists(location));
+                }
+            }
+        }
+        assertTrue(fileExistsInListing(location));
+        assertTrue(fileExists(location));
+        getFileSystem().deleteFile(location);
+    }
+
+    @Test
+    public void testLargeFileDoesNotOverwriteUntilClosed()
+            throws IOException
+    {
+        Location location = getRootLocation().appendPath("testFile4");
+        try (OutputStream outputStream = getFileSystem().newOutputFile(location).createOrOverwrite()) {
+            outputStream.write("test".getBytes(UTF_8));
+        }
+        TrinoInputFile inputFile = getFileSystem().newInputFile(location);
+        int fileSize = toIntExact(inputFile.length());
+        byte[] fileBytes = new byte[fileSize];
+        assertEquals(fileSize, readFile(inputFile, fileBytes));
+        assertEquals("test", new String(fileBytes, UTF_8));
+        byte[] bytes = new byte[8192];
+        for (int i = 0; i < bytes.length / 4; i++) {
+            //noinspection NumericCastThatLosesPrecision
+            bytes[i] = (byte) i;
+        }
+        try (OutputStream outputStream = getFileSystem().newOutputFile(location).createOrOverwrite()) {
+            // write a 32 mb file
+            for (int i = 0; i < divide(32 * 1024 * 1024, bytes.length, UNNECESSARY); i++) {
```

Comment:
Is 32 MB really necessary ? We do control the size of array being written (8K). Is there local caching that you're trying to avoid ?
There may be some issues on the CI when testing from a Github runner which is far from the region of the  GCS bucket.

Reply:
These were originally to test gcs filesystem: the write buffer under the hood is 16mb, so wanted to choose a size that would ensure we were exercising the case where the buffer flushes to storage. I can do 20mb - currently the 2 tests take 18 seconds and 14 seconds respectively, lmk if you'd like the size lowered.

Reply:
We could make it 17MB or 18MB